OpenRA.Mods.Common/Traits/Sound/AmbientSound.cs

```diff
@@ -22,43 +22,52 @@ class AmbientSoundInfo : UpgradableTraitInfo
 		[Desc("Interval between playing the sound (in ticks).")]
 		public readonly int Interval = 0;
 
-		public override object Create(ActorInitializer init) { return new AmbientSound(init.Self, this); }
+		public override object Create(ActorInitializer init) { return new AmbientSound(this); }
 	}
 
-	class AmbientSound : UpgradableTrait<AmbientSoundInfo>, ITick
+	class AmbientSound : UpgradableTrait<AmbientSoundInfo>, ITick, INotifyRemovedFromWorld
 	{
 		ISound currentSound;
 		bool wasDisabled = true;
 		int interval;
+		WPos cachedPosition;
 
-		public AmbientSound(Actor self, AmbientSoundInfo info)
+		public AmbientSound(AmbientSoundInfo info)
 			: base(info)
 		{
 			interval = info.Interval;
 		}
 
 		public void Tick(Actor self)
 		{
-			if (IsTraitDisabled)
+			if (IsTraitDisabled || !self.IsInWorld)
 			{
-				Game.Sound.StopSound(currentSound);
-				currentSound = null;
-				wasDisabled = true;
+				StopSound();
 				return;
 			}
 
-			if (wasDisabled && Info.Interval <= 0)
+			if (Info.Interval <= 0)
 			{
-				if (self.OccupiesSpace != null)
+				var moved = self.OccupiesSpace != null && cachedPosition != self.CenterPosition;
+				if (!wasDisabled && !moved)
+					return;
+
+				wasDisabled = false;
+
+				if (moved)
+				{
+					// Otherwise the sound never gets stopped when the actor is on the move
```

Comment:
This is kind of bogus.  If you cancel and restart the sound when the actor moves you will get serious s-s-s-stutt-stu-stutte-stuttering.